overlord/snapstate/autorefresh_gating.go

```diff
@@ -0,0 +1,201 @@
+		if inf.Hooks[gateAutoRefreshHookName] == nil {
+			delete(all, name)
+			continue
+		}
+
+		base := inf.Base
+		if base == "none" {
+			continue
+		}
+		if inf.Base == "" {
+			base = "core"
+		}
+		byBase[base] = append(byBase[base], inf.InstanceName())
+	}
+
+	affected := make(map[string]*affectedSnapInfo)
+
+	addAffected := func(snapName, affectedBy string, restart bool, base bool) {
+		affectedInfo := affected[snapName]
+		if affectedInfo == nil {
+			affectedInfo = &affectedSnapInfo{
+				AffectingSnaps: map[string]bool{},
+			}
+			affected[snapName] = affectedInfo
+		}
+		if restart {
+			affectedInfo.Restart = restart
+		}
+		if base {
+			affectedInfo.Base = base
+		}
+		affectedInfo.AffectingSnaps[affectedBy] = true
+	}
+
+	for _, up := range updates {
+		// on core system, affected by update of boot base
+		if bootBase != "" && up.InstanceName() == bootBase {
+			for _, snapSt := range all {
+				addAffected(snapSt.InstanceName(), up.InstanceName(), true, true)
+			}
+		}
+
+		// snaps that can trigger reboot
+		// XXX: gadget refresh doesn't always require reboot, refine this
+		if up.Type() == snap.TypeKernel || up.Type() == snap.TypeGadget {
+			for _, snapSt := range all {
+				addAffected(snapSt.InstanceName(), up.InstanceName(), true, false)
+			}
+			continue
+		}
+		if up.Type() == snap.TypeBase || up.SnapName() == "core" {
+			// affected by refresh of this base snap
+			for _, snapName := range byBase[up.InstanceName()] {
+				addAffected(snapName, up.InstanceName(), false, true)
+			}
+			continue
+		}
+
+		repo := ifacerepo.Get(st)
+
+		// consider slots provided by refreshed snap, but exclude core and snapd
+		// since they provide system-level slots that are generally not disrupted
+		// by snap updates.
+		if up.SnapType != snap.TypeSnapd && up.SnapName() != "core" {
+			for _, slotInfo := range up.Slots {
+				conns, err := repo.Connected(up.InstanceName(), slotInfo.Name)
+				if err != nil {
+					return nil, err
+				}
+				for _, cref := range conns {
+					// affected only if it wasn't optimized out above
+					if all[cref.PlugRef.Snap] != nil {
+						addAffected(cref.PlugRef.Snap, up.InstanceName(), false, false)
+					}
+				}
+			}
+		}
+
+		// consider plugs of the refreshed snap only if mount backend is involved
+		for _, plugInfo := range up.Plugs {
+			iface := repo.Interface(plugInfo.Interface)
+			if iface == nil {
+				return nil, fmt.Errorf("internal error: unknown interface %s", plugInfo.Interface)
+			}
+			if !usesMountBackend(iface) {
+				continue
+			}
+			conns, err := repo.Connected(up.InstanceName(), plugInfo.Name)
+			if err != nil {
+				return nil, err
+			}
+			for _, cref := range conns {
+				// affected only if it wasn't optimized out above
+				if all[cref.SlotRef.Snap] != nil {
+					addAffected(cref.SlotRef.Snap, up.InstanceName(), false, false)
+				}
+			}
+		}
+	}
+
+	return affected, nil
+}
+
+func usesMountBackend(iface interfaces.Interface) bool {
```

Comment:
I looked more, this is probably fine as is for now, at least until we make snap-update-ns less disruptive.  We don't have usages of MountPermanentPlug or MountConnectedSlot so for those it doesn't change much either way. If we wanted to do better we would need to attach probably more metadata to the interfaces themselves.